### PR TITLE
Fix several buffer over-reads

### PR DESCRIPTION
**Pull request by Algunenano**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Fix several buffer over-reads

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

2 commits so far:
* One fixes multiple buffer over-reads in aggregation states deserialization. This should be backported to all releases.
* The second one adds [[nodiscard]] to the read methods since they are clearly prone to be forgotten and leads to dangerous code. Backporting this shouldn't be necessary.

### CODE REVIEW

**Review by CheSema — APPROVED** [review]
Great! Thnx.

**Comment by Algunenano** [comment]
Seeing the CI of the other PR, this is going to make the fuzzers go crazy. The most common error I'm seeing is at Allocator::checkSize, which is a `LOGICAL_ERROR` but it can be triggered by the user. We could either manually check all the sizes or change it to `CANNOT_ALLOCATE_MEMORY`. Any thoughts?

**Comment by CheSema** [comment]
> 

I see that CANNOT_ALLOCATE_MEMORY is usually refers to the problem with system calls. Allocator::checkSize has a synthetic condition witch is seemed to be sensible. I think we have to fix all the issues with fuzzers manually by finding when it is asked for unreasonable memory size. I don't understand what we gain from LOGICAL_ERROR->CANNOT_ALLOCATE_MEMORY in Allocator::checkSize in general.
However, I agree, there are possibility that user could see LOGICAL_ERROR from Allocator::checkSize. If you have some examples lets discuss them. I'm convinced that cases require a local changes, not such general.

**Comment by Algunenano** [comment]
> I don't understand what we gain from changing LOGICAL_ERROR to CANNOT_ALLOCATE_MEMORY in Allocator::checkSize in general.

The idea is to not throw LOGICAL_ERROR on user input because that's considered fatal in non-release builds and the fuzzers will be hitting it like crazy (and stopping there). But I see that it's mostly `SequenceNextNode`, as the others use Int32 for the size and won't hit this check, so I'll add a safety check there instead similar to ` max_centroids_deserialize` in QuantileTDigest

**Comment by Algunenano** [comment]
The Tsan [failure seems unrelated to the changes](https://s3.amazonaws.com/clickhouse-test-reports/43159/0c30325aa44cd612d989dd185357943c6ccdec96/fuzzer_astfuzzertsan//report.html):

```
$ ack ThreadSanitizer\|== 
server (4).log
82:2022.11.11 16:00:55.232075 [ 151 ] {} <Information> RaftInstance:    === INIT RAFT SERVER ===
1429161:ThreadSanitizer: Out of memory. The process has exhausted 16384MB for size class 8192.
1429162:==151==ERROR: ThreadSanitizer: out of memory: allocator is trying to allocate 0x2000 bytes
1429271:==151==HINT: if you don't care about these errors you may set allocator_may_return_null=1
1429272:SUMMARY: ThreadSanitizer: out-of-memory (/workspace/clickhouse+0xba7d691) (BuildId: b1301d95db4ff4eaf1c24d076488c82ae6e8fa00) in malloc
1429273:==151==ERROR: ThreadSanitizer: out of memory: allocator is trying to allocate 0x2000 bytes
1429335:==151==HINT: if you don't care about these errors you may set allocator_may_return_null=1
1429336:SUMMARY: ThreadSanitizer: out-of-memory (/workspace/clickhouse+0xba7d9e3) (BuildId: b1301d95db4ff4eaf1c24d076488c82ae6e8fa00) in realloc
1429410:==151==ERROR: ThreadSanitizer: out of memory: allocator is trying to allocate 0x2000 bytes
1429467:==151==HINT: if you don't care about these errors you may set allocator_may_return_null=1
1429468:SUMMARY: ThreadSanitizer: out-of-memory (/workspace/clickhouse+0xba7d691) (BuildId: b1301d95db4ff4eaf1c24d076488c82ae6e8fa00) in malloc
1429495:==151==ERROR: ThreadSanitizer: out of memory: allocator is trying to allocate 0x2000 bytes
1429557:==151==HINT: if you don't care about these errors you may set allocator_may_return_null=1
1429558:SUMMARY: ThreadSanitizer: out-of-memory (/workspace/clickhouse+0xba7d9e3) (BuildId: b1301d95db4ff4eaf1c24d076488c82ae6e8fa00) in realloc
```

```
$ grep Fatal server\ \(4\).log --text
2022.11.11 16:13:39.891269 [ 424 ] {} <Fatal> BaseDaemon: ########################################
2022.11.11 16:13:39.891718 [ 424 ] {} <Fatal> BaseDaemon: (version 22.11.1.1, build id: B1301D95DB4FF4EAF1C24D076488C82AE6E8FA00) (from thread 377) (query_id: 644086f9-48a7-4a66-a16b-c422ce96a9e8) (query: SELECT k, (0.5, 10000000000., 10000000000.), groupArrayForEach(arr) FROM (SELECT number % NULL AS k, (NULL, 1000.0001, NULL), range(number) AS arr FROM system.numbers LIMIT 1024) GROUP BY k WITH ROLLUP WITH TOTALS ORDER BY k ASC NULLS FIRST) Received signal Unknown signal (-3)
2022.11.11 16:13:39.891882 [ 424 ] {} <Fatal> BaseDaemon: Sanitizer trap.
2022.11.11 16:13:39.892091 [ 424 ] {} <Fatal> BaseDaemon: Stack trace: 0x138d6133 0x13b5111a 0xba65216 0xba6a179 0xbadac06 0xbadb651 0xba7d69d 0x1386ce22 0x143f5e04 0x143f6208 0x143f313f 0x17627fb7 0x17625d80 0x17626824 0x1bb75d9b 0x1ba576f7 0x1b9984e3 0x1b99caf7 0x1dffd5f3 0x1dffad5f 0x1dd98e13 0x1dd8a0e1 0x1dd89423 0x1dd891fa 0x1dd9d479 0x1399bcb1 0x139a03d2 0xba7ed67 0x7f1016c00609 0x7f1016b25133
2022.11.11 16:13:39.964036 [ 425 ] {} <Fatal> BaseDaemon: ########################################
2022.11.11 16:13:39.964127 [ 425 ] {} <Fatal> BaseDaemon: (version 22.11.1.1, build id: B1301D95DB4FF4EAF1C24D076488C82AE6E8FA00) (from thread 188) (query_id: 06e0ed11-87b9-4af5-8334-8556f0fc33ed::202211_51_94_13) (query: ) Received signal Unknown signal (-3)
2022.11.11 16:13:39.964206 [ 425 ] {} <Fatal> BaseDaemon: Sanitizer trap.
2022.11.11 16:13:39.964329 [ 425 ] {} <Fatal> BaseDaemon: Stack trace: 0x138d6133 0x13b5111a 0xba65216 0xba6a179 0xbadac06 0xbadb800 0xba7d9f2 0x1386d4d2 0xc0edb34 0x1b54a16a 0x1b54751f 0x1b5465a3 0x1b5445a4 0x1ce8caf1 0x1ce8c8f4 0x1ce8c781 0x1b6b10e9 0x1d91d7c6 0x1d91c949 0x1d91ae36 0x1da46079 0x1d814fa4 0x1d821642 0x1d814e55 0x1d81a6a9 0x1db7a5b3 0x1d82f46f 0x1d8307c1 0x1d836cc2 0x1399ea11 0x139a2e98 0x139a2e02 0x1399bcb1 0x139a03d2 0xba7ed67 0x7f1016c00609 0x7f1016b25133
2022.11.11 16:13:39.966539 [ 424 ] {} <Fatal> BaseDaemon: 0.1. inlined from ./build_docker/../src/Common/StackTrace.cpp:332: StackTrace::tryCapture()
2022.11.11 16:13:39.966704 [ 424 ] {} <Fatal> BaseDaemon: 0. ./build_docker/../src/Common/StackTrace.cpp:293: StackTrace::StackTrace() @ 0x138d6133 in /workspace/clickhouse
2022.11.11 16:13:40.037793 [ 425 ] {} <Fatal> BaseDaemon: 0.1. inlined from ./build_docker/../src/Common/StackTrace.cpp:332: StackTrace::tryCapture()
2022.11.11 16:13:40.037857 [ 425 ] {} <Fatal> BaseDaemon: 0. ./build_docker/../src/Common/StackTrace.cpp:293: StackTrace::StackTrace() @ 0x138d6133 in /workspace/clickhouse
2022.11.11 16:13:40.121165 [ 424 ] {} <Fatal> BaseDaemon: 1. ./build_docker/../src/Daemon/BaseDaemon.cpp:415: sanitizerDeathCallback() @ 0x13b5111a in /workspace/clickhouse
2022.11.11 16:13:40.194603 [ 425 ] {} <Fatal> BaseDaemon: 1. ./build_docker/../src/Daemon/BaseDaemon.cpp:415: sanitizerDeathCallback() @ 0x13b5111a in /workspace/clickhouse
2022.11.11 16:13:50.449932 [ 425 ] {} <Fatal> BaseDaemon: 2. __sanitizer::Die() @ 0xba65216 in /workspace/clickhouse
2022.11.11 16:13:50.449932 [ 424 ] {} <Fatal> BaseDaemon: 2. __sanitizer::Die() @ 0xba65216 in /workspace/clickhouse
2022.11.11 16:13:53.891325 [ 426 ] {} <Fatal> BaseDaemon: ########################################
2022.11.11 16:13:53.891417 [ 426 ] {} <Fatal> BaseDaemon: (version 22.11.1.1, build id: B1301D95DB4FF4EAF1C24D076488C82AE6E8FA00) (from thread 194) (query_id: f0309520-4b63-4325-b40f-29072b546db0::202211_1_101_20) (query: ) Received signal Unknown signal (-3)
2022.11.11 16:13:53.891485 [ 426 ] {} <Fatal> BaseDaemon: Sanitizer trap.
2022.11.11 16:13:53.891609 [ 426 ] {} <Fatal> BaseDaemon: Stack trace: 0x138d6133 0x13b5111a 0xba65216 0xba6a179 0xbadac06 0xbadb651 0xba7d69d 0x1386ce22 0xbb0dd45 0xbb0dca6 0x1b6d281b 0x1b67b365 0x1d9d727c 0x1d9d69be 0x1d9ee65a 0x1dd73375 0x1dd72daa 0x1dd98e13 0x1dd8a0e1 0x1dd89729 0x1dd9e4c4 0x1dd9e7e1 0x1d814f49 0x1d821642 0x1d814e55 0x1d81a6a9 0x1db7a5b3 0x1d82f46f 0x1d8307c1 0x1d836cc2 0x1399ea11 0x139a2e98 0x139a2e02 0x1399bcb1 0x139a03d2 0xba7ed67 0x7f1016c00609 0x7f1016b25133
2022.11.11 16:13:53.963561 [ 426 ] {} <Fatal> BaseDaemon: 0.1. inlined from ./build_docker/../src/Common/StackTrace.cpp:332: StackTrace::tryCapture()
2022.11.11 16:13:53.963633 [ 426 ] {} <Fatal> BaseDaemon: 0. ./build_docker/../src/Common/StackTrace.cpp:293: StackTrace::StackTrace() @ 0x138d6133 in /workspace/clickhouse
2022.11.11 16:13:54.112346 [ 426 ] {} <Fatal> BaseDaemon: 1. ./build_docker/../src/Daemon/BaseDaemon.cpp:415: sanitizerDeathCallback() @ 0x13b5111a in /workspace/clickhouse
2022.11.11 16:13:54.684713 [ 427 ] {} <Fatal> BaseDaemon: ########################################
2022.11.11 16:13:54.684797 [ 427 ] {} <Fatal> BaseDaemon: (version 22.11.1.1, build id: B1301D95DB4FF4EAF1C24D076488C82AE6E8FA00) (from thread 193) (query_id: 3c5b475c-9e50-4748-a17b-06375605e4b4::202211_1_111_22) (query: ) Received signal Unknown signal (-3)
2022.11.11 16:13:54.684881 [ 427 ] {} <Fatal> BaseDaemon: Sanitizer trap.
2022.11.11 16:13:54.685009 [ 427 ] {} <Fatal> BaseDaemon: Stack trace: 0x138d6133 0x13b5111a 0xba65216 0xba6a179 0xbadac06 0xbadb800 0xba7d9f2 0x1386d4d2 0xc0edb34 0x1b54a16a 0x1b54751f 0x1b5465a3 0x1b5445a4 0x1ce87bd7 0x1b6b3bb4 0x1d9d727c 0x1d9d69be 0x1d9ee65a 0x1dd73375 0x1dd72daa 0x1dd98e13 0x1dd8a0e1 0x1dd89729 0x1dd9e4c4 0x1dd9e7e1 0x1d814f49 0x1d821642 0x1d814e55 0x1d81a6a9 0x1db7a5b3 0x1d82f46f 0x1d8307c1 0x1d836cc2 0x1399ea11 0x139a2e98 0x139a2e02 0x1399bcb1 0x139a03d2 0xba7ed67 0x7f1016c00609 0x7f1016b25133
2022.11.11 16:13:54.757422 [ 427 ] {} <Fatal> BaseDaemon: 0.1. inlined from ./build_docker/../src/Common/StackTrace.cpp:332: StackTrace::tryCapture()
2022.11.11 16:13:54.757506 [ 427 ] {} <Fatal> BaseDaemon: 0. ./build_docker/../src/Common/StackTrace.cpp:293: StackTrace::StackTrace() @ 0x138d6133 in /workspace/clickhouse
2022.11.11 16:13:54.906808 [ 427 ] {} <Fatal> BaseDaemon: 1. ./build_docker/../src/Daemon/BaseDaemon.cpp:415: sanitizerDeathCallback() @ 0x13b5111a in /workspace/clickhouse
2022.11.11 16:13:56.624085 [ 425 ] {} <Fatal> BaseDaemon: 3. ? @ 0xba6a179 in /workspace/clickhouse
2022.11.11 16:13:56.624085 [ 424 ] {} <Fatal> BaseDaemon: 3. ? @ 0xba6a179 in /workspace/clickhouse
2022.11.11 16:13:59.576061 [ 426 ] {} <Fatal> BaseDaemon: 2. __sanitizer::Die() @ 0xba65216 in /workspace/clickhouse
```

No more errors so far (but a few pending tasks)

**Comment by Algunenano** [comment]
Failures:
* `AST Fuzzer (tsan)`: TSAN OOM
* `ClickHouse special build check `. Failed in `relwithdebuginfo`. I can't see the logs but it seems the connection to docker failed.
* `Mergeable Check`. Depends on the build passing.
* `Stress test (msan) `: OOM

I'll update with master, but these seems ok to me.

File: src/AggregateFunctions/AggregateFunctionNull.h [hunk 1]
```diff
@@ -153,7 +153,9 @@ class AggregateFunctionNullBase : public IAggregateFunctionHelper<Derived>
 
     void deserialize(AggregateDataPtr __restrict place, ReadBuffer & buf, std::optional<size_t> version, Arena * arena) const override
     {
-        bool flag = true;
+        /// When deserializing a bool it might trigger UBSAN if the input is not 0 or 1, so it's better to use a Int8
+        static_assert(sizeof(bool) == sizeof(Int8));
```
Comment (Algunenano) [inline src/AggregateFunctions/AggregateFunctionNull.h:157]: Similar sanitizer alerts are popping up in other places, so I'm going to change the readBinary(bool) implementation instead

**Comment by Algunenano** [comment]
Failures:
* AST fuzzer (msan): -> https://github.com/ClickHouse/ClickHouse/issues/42787
* AST fuzzer (ubsan): Issue in already existing code detected by ubsan. I'll try to address it here too.
* Stress test (msan): OOM.

**Comment by CheSema** [comment]
AST fuzzer (ubsan)
https://github.com/ClickHouse/ClickHouse/issues/43247

**Comment by Algunenano** [comment]
> AST fuzzer (ubsan) https://github.com/ClickHouse/ClickHouse/issues/43247

I'll send a PR for that. The fix is simple but I want to fix also the undefined behaviour in Arena.h too

**Comment by Algunenano** [comment]
This PR contained fixes for several critical vulnerabilities. Would it be possible to backport it to supported releases, please?

I can send manual PRs, but those won't run in the CI

**Pull request merged**